lib/versioning/regex/index.ts

```diff
@@ -0,0 +1,149 @@
+
+interface Version {
+  // major, minor, and patch act in the standard semver fashion, but without
+  // correctness requirements: if any one or two is omitted, they are treated
+  // as zero. At least one must be present
+  major: number;
+  minor: number;
+  patch: number;
+  // prereleases are treated in the standard semver manner, if present
+  prerelease: string;
+  // architectures, if present, are treated as a compatibility layer: we will
+  // never try to update to a version with a different architecture. Other than
+  // for juding compatibility (exact string match), the architecture is ignored
+  architecture: string;
+}
+
+// config is expected to be overridden by a user-specified RegExp value
+// sample values:
+//
+// * emulates the "semver" configuration:
+//   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<prerelease>.*))?$')
+// * emulates the "docker" configuration:
+//   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<architecture>.*))?$')
+// * matches the versoining scheme used by the Python images on DockerHub:
+//   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(?<prerelease>[^.-]+)?(-(?<architecture>.*))?$');
+let config = null;
+
+// convenience method for passing a string into a Version given current config.
+function parse(version: string): Version | null {
+  const match = version.match(config);
+  if (match === null) {
+    return null;
+  }
+
+  const groups = match.groups;
+  return {
+    major: typeof groups.major === 'undefined' ? 0 : Number(groups.major),
+    minor: typeof groups.minor === 'undefined' ? 0 : Number(groups.minor),
+    patch: typeof groups.patch === 'undefined' ? 0 : Number(groups.patch),
+    prerelease: groups.prerelease,
+    architecture: groups.architecture,
+  };
+}
+
+// convenience method for passing a Version object into any semver.* method.
+function asSemver(version: Version): string {
+  let vstring = `${version.major}.${version.minor}.${version.patch}`;
+  if (typeof version.prerelease !== 'undefined') {
+    vstring += `-${version.prerelease}`;
+  }
+  return vstring;
+}
+
+function configure(new_config: string): void {
+  // without at least one of {major, minor, patch} specified in the regex,
+  // this versioner will not work properly
+  // TODO: there's got to be a better way to do this
+  if (
+    !new_config.includes('<major>') &&
+    !new_config.includes('<minor>') &&
+    !new_config.includes('<patch>')
+  ) {
+    logger.fatal(
+      { new_config },
+      'Invalid versionConfig for versionScheme regex'
+    );
+    return;
+  }
+
+  // TODO: should we validate the user has not added extra unsupported
+  // capture groups?
+
+  config = RegExp(new_config);
+}
+
+const isValid = (version: string): string | boolean | null =>
+  parse(version) !== null;
+const isCompatible = (version: string, range: string): boolean =>
+  parse(version).architecture === parse(range).architecture;
```

Comment:
@rarkins should we check for possible null refs?

Reply:
I was looking into this a while back, especially in response to https://github.com/renovatebot/renovate/pull/4273#discussion_r312353683, and went looking at the other schemes to see how its been done before -- as far as I could tell, most of the other schemes currently obey this format, where generally most `VersioningApi` methods will raise an exception on un-parseable version schemes (in anything but the `isValid` method(s)). Figured it might be best to keep with the current style, for now?